src/lib/AudioUltra/Media/MediaLoader.ts

```diff
@@ -45,29 +67,48 @@ export class MediaLoader extends Destructable {
       return Promise.resolve(null);
     }
 
-    const audio = this.createAnalyzer(options);
-    const xhr = await this.performRequest(this.options.src);
+    this.decoderPromise = new Promise((resolve) => {
+      this.decoderResolve = resolve;
+    });
 
-    if (xhr.status === 200 && xhr.response) {
-      const playAudio = (buffer: AudioBuffer) => {
-        this.duration = buffer.duration;
-        this.sampleRate = audio.sampleRate ?? buffer.sampleRate;
-        this.loaded = true;
-        audio.buffer = buffer;
-        audio.connect();
-        return audio;
-      };
+    this.createAnalyzer({
+      ...options,
+      src: this.options.src,
+    });
 
+    if (!this.audio) {
+      return Promise.resolve(null);
+    }
+
+    if (await this.audio.sourceDecoded()) {
+      this.duration = this.audio.duration;
+      this.decoderResolve?.();
+      return this.audio;
+    }
+
+    const req = await this.performRequest(this.options.src);
+
+    if (req) {
       try {
-        if (!audio.context) {
+        if (!this.audio) {
           return Promise.resolve(null);
         }
 
-        return this.decodeAudioData(xhr.response).then((buffer) => {
-          if (buffer) {
-            return playAudio(buffer);
-          }
-          return null;
+        const decodingPromise = this.decodeAudioData(req);
+
+        if (this.audio) {
```

Comment:
is this block intented to go in parallel? if yes it's better to exlpicitely say this here

Reply:
but given that we already passed `if (!this.audio)` check this one is redundant